File: src/epay_tool/core/pipeline.py
"""Product enrichment and publishing pipeline.

Enrichment per product:

  1. Barcode -> eBay Catalog API (exact match: title, brand, aspects, EPID)
  2. No match + photos -> AI identification (with questions on uncertainty)
  3. Category + required aspects via Taxonomy API
  4. Market price research via Browse API -> price suggestion
  5. Missing title/description -> AI copywriting
  6. Deterministic quality check -> READY or NEEDS_INFO

Each step is defensive: a failing step records the problem and moves the
product to NEEDS_INFO/FAILED instead of crashing a 100-item batch.
"""

from __future__ import annotations

import logging
from pathlib import Path

from epay_tool.ai.provider import AIProvider
from epay_tool.config import Settings
from epay_tool.core import pricing, quality
from epay_tool.db.models import DataSource, Photo, Product, ProductState
from epay_tool.db.session import get_session
from epay_tool.ebay import browse, catalog, inventory, media, taxonomy
from epay_tool.ebay.client import EbayClient

log = logging.getLogger(__name__)


class Pipeline:
    def __init__(self, settings: Settings, client: EbayClient, ai: AIProvider):
        self.settings = settings
        self.client = client
        self.ai = ai

    # ---------------- enrichment ----------------

    def enrich(self, product_id: int) -> None:
        with get_session() as session:
            product = session.get(Product, product_id)
            if product is None:
                return
            product.state = ProductState.ENRICHING
            product.last_error = None
            session.commit()

            try:
                self._enrich_inner(product)
                issues = self._final_check(product)
                product.state = (
                    ProductState.NEEDS_INFO
                    if quality.has_blockers(issues) or product.ai_questions
                    else ProductState.READY
                )
            except Exception as exc:  # keep the batch alive
                log.exception("Enrichment failed for product %s", product_id)
                product.state = ProductState.FAILED
                product.last_error = str(exc)
            session.commit()

    def _enrich_inner(self, product: Product) -> None:
        if not product.condition:
            product.condition = self.settings.default_condition

        # 1. Catalog lookup by barcode
        if product.gtin and not product.title:
            match = catalog.find_by_gtin(self.client, product.gtin)
            if match:
                self._apply_catalog(product, match)

        # 2. AI fallback from photos
        if not product.title and product.photos:
            self._ai_identify(product)

        # 3. Category + required aspects
        if not product.category_id and product.title:
            suggestion = taxonomy.suggest_category(self.client, product.title)
            if suggestion:
                product.category_id = suggestion[0]

        # 4. Price research
        if product.price is None:
            stats = browse.price_research(
                self.client, gtin=product.gtin, query=product.title
            )
            if stats:
                product.price_stats = stats.to_dict()
                product.price = pricing.suggest_price(
                    stats, self.settings.undercut_percent
                )

        # 5. Copywriting
        if product.title and not product.description:
            facts = {
                "title": product.title,
                "brand": product.brand,
                "mpn": product.mpn,
                "condition": product.condition,
                "aspects": product.aspects,
                "notes": product.user_notes,
            }
            copy = self.ai.write_copy({k: v for k, v in facts.items() if v})
            product.title = copy.title[:80]
            product.description = copy.description_html

    def _apply_catalog(self, product: Product, match: catalog.CatalogMatch) -> None:
        product.source = DataSource.CATALOG
        product.epid = match.epid
        product.title = match.title
        product.brand = match.brand or product.brand
        product.mpn = match.mpn or product.mpn
        product.aspects = {**match.aspects, **(product.aspects or {})}
        if match.category_ids and not product.category_id:
            product.category_id = match.category_ids[0]

        # Catalog images are already eBay-hosted, so they need no upload and
        # make a barcode-only scan directly listable for new goods.
        if not product.photos:
            product.photos = [
                Photo(path="", ebay_url=url, position=i)
                for i, url in enumerate(match.image_urls[:12])
            ]

    def _ai_identify(self, product: Product) -> None:
        known = {
            k: v
            for k, v in {"gtin": product.gtin, "brand": product.brand}.items()
            if v
        }
        result = self.ai.identify_product(
            [Path(p.path) for p in product.photos],
            known=known,
            notes=product.user_notes,
        )
        product.source = DataSource.AI
        fields = result.confident_fields()
        if "product_name" in fields:
            name = fields["product_name"]
            brand = fields.get("brand", "")
            model = fields.get("model", "")
            parts = [p for p in (brand, name if name != brand else "", model) if p]
            product.title = " ".join(dict.fromkeys(" ".join(parts).split()))[:80]
        product.brand = fields.get("brand") or product.brand
        # Only confident values may reach the listing; an uncertain model guess
        # becomes a question to the user instead of a fabricated part number.
        product.mpn = product.mpn or fields.get("model")
        confident_aspects = {
            name: [guess.value]
            for name, guess in result.aspects.items()
            if guess.confidence >= 0.8
        }
        product.aspects = {**confident_aspects, **(product.aspects or {})}
        product.ai_questions = result.questions or None
        product.ai_suggestions = result.photo_suggestions or None
        product.ai_confidence = {
            name: guess.confidence
            for name, guess in {
                "product_name": result.product_name,
                "brand": result.brand,
                "model": result.model,
                **result.aspects,
            }.items()
            if guess
        }

    def _final_check(self, product: Product) -> list[quality.Issue]:
        required = []
        if product.category_id:
            try:
                required = taxonomy.aspects_for_category(
                    self.client, product.category_id
                )
            except Exception:
                log.warning("Could not load required aspects", exc_info=True)
        return quality.check_draft(product, required)

    # ---------------- publishing ----------------

    def publish(
        self,
        product_id: int,
        *,
        fulfillment_policy_id: str,
        payment_policy_id: str,
        return_policy_id: str,
    ) -> None:
        """Upload photos, create the offer and publish it - the 'one click'."""
        with get_session() as session:
            product = session.get(Product, product_id)
            if product is None:
                return
            product.state = ProductState.POSTING
            session.commit()
            try:
                image_urls = []
                for photo in product.photos:
                    if not photo.ebay_url:
                        photo.ebay_url = media.upload_image(self.client, Path(photo.path))
                    image_urls.append(photo.ebay_url)

                inventory.upsert_inventory_item(self.client, product, image_urls)
                product.offer_id = inventory.create_or_update_offer(
                    self.client,
                    product,
                    fulfillment_policy_id=fulfillment_policy_id,
                    payment_policy_id=payment_policy_id,
                    return_policy_id=return_policy_id,
                )
                product.listing_id = inventory.publish_offer(self.client, product.offer_id)
                product.state = ProductState.POSTED
            except Exception as exc:
                log.exception("Publishing failed for product %s", product_id)
                product.state = ProductState.FAILED
                product.last_error = str(exc)
            session.commit()

File: src/main/core/pipeline.ts
/**
 * Product enrichment and publishing pipeline.
 *
 * Enrichment per product:
 *
 *   1. Barcode -> eBay Catalog API (exact match: title, brand, aspects, EPID)
 *   2. No match + photos -> AI identification (with questions on uncertainty)
 *   3. Category + required aspects via Taxonomy API
 *   4. Market price research via Browse API -> price suggestion
 *   5. Missing title/description -> AI copywriting
 *   6. Deterministic quality check -> READY or NEEDS_INFO
 *
 * Each step is defensive: a failing product records its error and moves
 * to FAILED instead of taking a 100-item batch down with it.
 */

import type { AIProvider } from '@main/ai/provider'
import { confidenceMap, confidentAspects, confidentFields } from '@main/ai/schemas'
import type { AppConfig } from '@main/config'
import type { ProductRepository, ProductUpdate } from '@main/db/products'
import { priceResearch } from '@main/ebay/browse'
import type { CatalogMatch } from '@main/ebay/catalog'
import type { Capability } from '@main/ebay/capabilities'
import { findByGtin } from '@main/ebay/catalog'
import { asCapabilityError, type EbayClient } from '@main/ebay/client'
import type { TokenSource } from '@main/ebay/client'
import { createOrUpdateOffer, publishOffer, upsertInventoryItem } from '@main/ebay/inventory'
import { uploadImage } from '@main/ebay/media'
import type { AspectRequirement } from '@main/ebay/taxonomy'
import { aspectsForCategory, suggestCategory } from '@main/ebay/taxonomy'
import { DataSource, ProductState, type Product } from '@shared/types'

import { suggestPrice } from './pricing'
import { checkDraft, hasBlockers } from './quality'

const MAX_STOCK_PHOTOS = 12
const MAX_TITLE_LENGTH = 80

export interface PipelineDeps {
  config: AppConfig
  client: EbayClient
  auth: TokenSource
  ai: AIProvider
  repo: ProductRepository
  logger?: Pick<Console, 'warn' | 'error'>
}

export class Pipeline {
  private readonly config: AppConfig
  private readonly client: EbayClient
  private readonly auth: TokenSource
  private readonly ai: AIProvider
  private readonly repo: ProductRepository
  private readonly logger: Pick<Console, 'warn' | 'error'>
  private readonly unavailable = new Set<Capability>()

  constructor(deps: PipelineDeps) {
    this.config = deps.config
    this.client = deps.client
    this.auth = deps.auth
    this.ai = deps.ai
    this.repo = deps.repo
    this.logger = deps.logger ?? console
  }

  // ---------------- enrichment ----------------

  async enrich(productId: number): Promise<void> {
    const product = this.repo.get(productId)
    if (!product) return

    this.repo.setState(productId, ProductState.Enriching, null)

    try {
      const changes = await this.buildEnrichment(product)
      this.repo.update(productId, changes)

      const updated = this.repo.get(productId)
      if (!updated) return

      const issues = await this.finalCheck(updated)
      const needsInfo = hasBlockers(issues) || Boolean(updated.aiQuestions?.length)
      this.repo.setState(productId, needsInfo ? ProductState.NeedsInfo : ProductState.Ready, null)
    } catch (error) {
      this.logger.error(`Enrichment failed for product ${productId}`, error)
      this.repo.setState(productId, ProductState.Failed, describeError(error))
    }
  }

  private async buildEnrichment(product: Product): Promise<ProductUpdate> {
    const changes: ProductUpdate = {}
    const current = (): Product => ({ ...product, ...changes } as Product)

    if (!product.condition) {
      changes.condition = this.config.defaultCondition
    }

    // 1. Catalog lookup by barcode. A missing entitlement is not a
    //    product failure - it just means this step cannot contribute.
    const gtin = product.gtin
    if (gtin && !product.title) {
      const match = await this.optional('catalog', () => findByGtin(this.client, gtin))
      if (match) this.applyCatalog(product, match, changes)
    }

    // 2. AI fallback from photos.
    if (!changes.title && !product.title && product.photos.length > 0) {
      await this.applyAiIdentification(product, changes)
    }

    // 3. Category + required aspects.
    const title = changes.title ?? product.title
    if (!product.categoryId && !changes.categoryId && title) {
      const suggestion = await this.optional('taxonomy', () =>
        suggestCategory(this.client, title),
      )
      if (suggestion) changes.categoryId = suggestion[0]
    }

    // 4. Price research.
    if (product.price === null) {
      const stats = await this.optional('browse', () =>
        priceResearch(this.client, { gtin: product.gtin, query: title }),
      )
      if (stats) {
        changes.priceStats = stats
        changes.price = suggestPrice(stats, this.config.undercutPercent)
      }
    }

    // 5. Copywriting.
    if (title && !product.description) {
      const source = current()
      const facts = pruneEmpty({
        title,
        brand: source.brand,
        mpn: source.mpn,
        condition: source.condition,
        aspects: source.aspects,
        notes: source.userNotes,
      })
      const copy = await this.ai.writeCopy(facts)
      changes.description = copy.descriptionHtml

      // A catalog title is the exact wording eBay holds for this EPID.
      // Rewriting it would contradict the catalog entry the listing is
      // linked to, so the AI only retitles products it identified itself.
      if (changes.source !== DataSource.Catalog && product.source !== DataSource.Catalog) {
        changes.title = copy.title.slice(0, MAX_TITLE_LENGTH)
      }
    }

    return changes
  }

  /**
   * Runs an enrichment step that the application may not be entitled to
   * call. A missing entitlement yields null and is recorded once for the
   * whole session; every other error still propagates and fails the
   * product, because that is a genuine fault worth showing.
   */
  private async optional<T>(capability: Capability, step: () => Promise<T>): Promise<T | null> {
    if (this.unavailable.has(capability)) return null
    try {
      return await step()
    } catch (error) {
      const unavailable = asCapabilityError(error, capability)
      if (!unavailable) throw error

      if (!this.unavailable.has(capability)) {
        this.unavailable.add(capability)
        this.logger.warn(`eBay capability unavailable: ${capability}`, unavailable.message)
      }
      return null
    }
  }

  /** Capabilities eBay has refused during this session. */
  get missingCapabilities(): Capability[] {
    return [...this.unavailable]
  }

  private applyCatalog(product: Product, match: CatalogMatch, changes: ProductUpdate): void {
    changes.source = DataSource.Catalog
    changes.epid = match.epid
    changes.title = match.title
    changes.brand = match.brand ?? product.brand
    changes.mpn = match.mpn ?? product.mpn
    changes.aspects = { ...match.aspects, ...(product.aspects ?? {}) }
    if (match.categoryIds.length > 0 && !product.categoryId) {
      changes.categoryId = match.categoryIds[0] as string
    }

    // Catalog images are already eBay-hosted, so they need no upload and
    // make a barcode-only scan directly listable for new goods.
    if (product.photos.length === 0 && match.imageUrls.length > 0) {
      this.repo.addStockPhotos(product.id, match.imageUrls.slice(0, MAX_STOCK_PHOTOS))
    }
  }

  private async applyAiIdentification(product: Product, changes: ProductUpdate): Promise<void> {
    const known = pruneEmpty({ gtin: product.gtin, brand: product.brand }) as Record<string, string>

    const identification = await this.ai.identifyProduct({
      photoPaths: product.photos.map((photo) => photo.path).filter(Boolean),
      known,
      notes: product.userNotes,
    })

    changes.source = DataSource.Ai

    const fields = confidentFields(identification)
    if (fields.productName) {
      changes.title = buildTitle(fields).slice(0, MAX_TITLE_LENGTH)
    }
    changes.brand = fields.brand ?? product.brand
    // Only confident values may reach the listing; an uncertain model guess
    // becomes a question to the user instead of a fabricated part number.
    changes.mpn = product.mpn ?? fields.model ?? null
    changes.aspects = { ...confidentAspects(identification), ...(product.aspects ?? {}) }
    changes.aiQuestions = identification.questions.length ? identification.questions : null
    changes.aiSuggestions = identification.photoSuggestions.length
      ? identification.photoSuggestions
      : null
    changes.aiConfidence = confidenceMap(identification)
  }

  private async finalCheck(product: Product): Promise<ReturnType<typeof checkDraft>> {
    let required: AspectRequirement[] = []
    if (product.categoryId) {
      try {
        required = await aspectsForCategory(this.client, product.categoryId)
      } catch (error) {
        // A taxonomy hiccup must not fail the product; it only means the
        // required-aspect check is skipped for this pass.
        this.logger.warn('Could not load required aspects', error)
      }
    }
    return checkDraft(product, required)
  }

  // ---------------- publishing ----------------

  /** Upload photos, create the offer and publish it - the "one click". */
  async publish(
    productId: number,
    policies: { fulfillment: string; payment: string; return: string },
  ): Promise<void> {
    const product = this.repo.get(productId)
    if (!product) return

    this.repo.setState(productId, ProductState.Posting, null)

    try {
      const imageUrls: string[] = []
      for (const photo of product.photos) {
        if (photo.ebayUrl) {
          imageUrls.push(photo.ebayUrl)
          continue
        }
        const url = await uploadImage(this.config, this.auth, photo.path)
        this.repo.setPhotoUrl(photo.id, url)
        imageUrls.push(url)
      }

      await upsertInventoryItem(this.client, product, imageUrls)
      const offerId = await createOrUpdateOffer(this.client, product, {
        fulfillmentPolicyId: policies.fulfillment,
        paymentPolicyId: policies.payment,
        returnPolicyId: policies.return,
      })
      const listingId = await publishOffer(this.client, offerId)

      this.repo.update(productId, {
        offerId,
        listingId,
        state: ProductState.Posted,
        lastError: null,
      })
    } catch (error) {
      this.logger.error(`Publishing failed for product ${productId}`, error)
      this.repo.setState(productId, ProductState.Failed, describeError(error))
    }
  }
}

function buildTitle(fields: { productName?: string; brand?: string; model?: string }): string {
  const parts = [fields.brand, fields.productName, fields.model].filter(Boolean) as string[]
  // Drop duplicate words, e.g. brand repeated inside the product name.
  const seen = new Set<string>()
  const words: string[] = []
  for (const word of parts.join(' ').split(/\s+/)) {
    const key = word.toLowerCase()
    if (word && !seen.has(key)) {
      seen.add(key)
      words.push(word)
    }
  }
  return words.join(' ')
}

function pruneEmpty(record: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(record).filter(([, value]) => value !== null && value !== undefined && value !== ''),
  )
}

function describeError(error: unknown): string {
  return error instanceof Error ? error.message : String(error)
}
